example/main.mo

```diff
@@ -0,0 +1,255 @@
+    };
+    #Err : {
+      #CallLedgerError : Text;
+      #NotAvailable : Text;
+    };
+  };
+
+  type WithdrawResult = {
+    #Ok : {
+      txid : Nat;
+      amount : Nat;
+    };
+    #Err : {
+      #CallLedgerError : Text;
+      #InsufficientCredit;
+      #AmountBelowMinimum;
+    };
+  };
+
+  let JOURNAL_SIZE = 1024;
+
+  var initialized : Bool = false;
+  var assets : Vec.Vector<AssetInfo> = Vec.new();
+
+  private func assertInitialized() = if (not initialized) {
+    Prim.trap("Not initialized");
+  };
+
+  public shared func init() : async () {
+    assert not initialized;
+    assets := Vec.map<StableAssetInfo, AssetInfo>(
+      assetsData,
+      func(x) {
+        let r = (actor (Principal.toText(x.ledgerPrincipal)) : ICRC1.ICRC1Ledger)
+        |> {
+          balance_of = _.icrc1_balance_of;
+          fee = _.icrc1_fee;
+          transfer = _.icrc1_transfer;
+          transfer_from = _.icrc2_transfer_from;
+        }
+        |> {
+          ledgerPrincipal = x.ledgerPrincipal;
+          handler = TokenHandler.TokenHandler(_, Principal.fromActor(self), JOURNAL_SIZE, 0, true);
+        };
+        r.handler.unshare(x.handler);
+        r;
+      },
+    );
+    initialized := true;
+  };
+
+  public shared query func principalToSubaccount(p : Principal) : async ?Blob = async ?TokenHandler.toSubaccount(p);
+
+  public shared query func icrcX_supported_tokens() : async [Principal] {
+    assertInitialized();
+    Array.tabulate<Principal>(
+      Vec.size(assets),
+      func(i) = Vec.get(assets, i).ledgerPrincipal,
+    );
+  };
+
+  public shared query func icrcX_token_info(token : Principal) : async TokenInfo {
+    assertInitialized();
+    for ((assetInfo, i) in Vec.items(assets)) {
+      if (Principal.equal(assetInfo.ledgerPrincipal, token)) {
+        return {
+          min_deposit = assetInfo.handler.minimum(#deposit);
+          min_withdrawal = assetInfo.handler.minimum(#withdrawal);
+          deposit_fee = assetInfo.handler.fee(#deposit);
+          withdrawal_fee = assetInfo.handler.fee(#withdrawal);
+        };
+      };
+    };
+    throw Error.reject("Unknown token");
+  };
+
+  public shared query ({ caller }) func icrcX_credit(token : Principal) : async Int {
+    assertInitialized();
+    switch (getAssetInfo(token)) {
+      case (?info) info.handler.userCredit(caller);
+      case (_) 0;
+    };
+  };
+
+  public shared query ({ caller }) func icrcX_all_credits() : async [(Principal, Int)] {
+    assertInitialized();
+    let res : Vec.Vector<(Principal, Int)> = Vec.new();
+    for (assetInfo in Vec.vals(assets)) {
+      let credit = assetInfo.handler.userCredit(caller);
+      if (credit != 0) {
+        Vec.add(res, (assetInfo.ledgerPrincipal, credit));
+      };
+    };
+    Vec.toArray(res);
+  };
+
+  public shared query ({ caller }) func icrcX_trackedDeposit(token : Principal) : async {
+    #Ok : Nat;
+    #Err : { #NotAvailable : Text };
+  } {
+    assertInitialized();
```

Comment:
The icrcX spec says to throw an error (i.e. `throw Error.reject("..")`) in case the token isn't known. It does not return a Result type, i.e. no #Ok/#Err variant.